lib/wallet/coinselector.js

```diff
@@ -0,0 +1,882 @@
+/*!
+ * coinselector.js - coin selector class for bcoin
+ * Copyright (c) 2014-2015, Fedor Indutny (MIT License)
+ * Copyright (c) 2014-2017, Christopher Jeffrey (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const {BufferMap} = require('buffer-map');
+const Output = require('../primitives/output');
+const Coin = require('../primitives/coin');
+const Outpoint = require('../primitives/outpoint');
+const Address = require('../primitives/address');
+const consensus = require('../protocol/consensus');
+const policy = require('../protocol/policy');
+const Amount = require('../btc/amount');
+
+/**
+ * CoinPointer
+ */
+
+class CoinPointer {
+  /**
+   * Create a credit.
+   * @constructor
+   * @param {Number} spendingSize
+   * @param {Number} effectiveValue
+   * @param {number} index
+   */
+
+  constructor(spendingSize, effectiveValue, index) {
+    this.spendingSize = spendingSize;
+    this.effectiveValue = effectiveValue;
+    this.index = index;
+  }
+}
+
+/**
+ * Coin Selector
+ * @alias module:primitives.CoinSelector
+ */
+
+class CoinSelector {
+  /**
+   * Create a coin selector.
+   * @constructor
+   * @param {TX} tx
+   * @param {Object?} options
+   */
+
+  constructor(tx, options) {
+    this.tx = tx.clone();
+    this.coins = [];
+    this.coinPointers = [];
+    this.outputValue = 0;
+    this.index = 0;
```

Comment:
There's still some class properties left over from when there was only one algorithm in here. I'd still like to see if `this.index` is necessary and I wonder if we even still need `this.fee` and `this.change` etc to be "global" in the coin selector

Reply:
Cannot remove `this.fee` and `this.change` as they are required here.
https://github.com/bcoin-org/bcoin/blob/aba6841e43546e8a485e96dc0019d1e788eab2ee/lib/primitives/mtx.js#L1251-L1265

Reply:
We can remove `this.index` but that would require changes to `selectEstimate()`. Should it be in this PR?